### ISSUE TITLE
some strange results in ultralytics

### ISSUE THREAD

**Issue by luoshiyong** [ultralytics/ultralytics#19307]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Because the instance branch in yoloseg has a operate "matrix mutiple", the infer process is slow in mobile device. so i want modified the instance branch to semantic branch, just make the protos nm=1 to predict the segmap,  detail below:`def single_mask_loss(self,
        gt_mask: torch.Tensor, pred: torch.Tensor, proto: torch.Tensor, xyxy: torch.Tensor, area: torch.Tensor
    ) -> torch.Tensor:
        """
        Compute the instance segmentation loss for a single image.

        Args:
            gt_mask (torch.Tensor): Ground truth mask of shape (n, H, W), where n is the number of objects.
            pred (torch.Tensor): Predicted mask coefficients of shape (n, 32).
            proto (torch.Tensor): Prototype masks of shape (32, H, W).
            xyxy (torch.Tensor): Ground truth bounding boxes in xyxy format, normalized to [0, 1], of shape (n, 4).
            area (torch.Tensor): Area of each ground truth bounding box of shape (n,).

        Returns:
            (torch.Tensor): The calculated mask loss for a single image.

        Notes:
            The function uses the equation pred_mask = torch.einsum('in,nhw->ihw', pred, proto) to produce the
            predicted masks from the prototype masks and predicted mask coefficients.
        """
        pred_mask = F.interpolate(proto[None], (proto.shape[1]*4, proto.shape[2]*4), mode="bilinear")[0] # [1,704,704]
        merged_mask = torch.any(gt_mask > 0, dim=0).unsqueeze(0).float()  # 转换为 float 类型[1,704,704]
        train_gt0 = merged_mask.squeeze(0).detach().cpu().numpy()
        train_gt0[train_gt0 == 1] = 255
        cv2.imwrite("gtseg.jpg",train_gt0)
        loss = F.binary_cross_entropy_with_logits(pred_mask, merged_mask, reduction="none")
        return (crop_mask(loss, xyxy).mean(dim=(1, 2)) / area).sum()`
but a strange thing appeared.  when I finished my trainning, the box ap is good but the seg ap is low, the reason i know is that the target is small object so the position can be find accurate but the shape of target is duffcult to predict, so the iou of seg pred is low, which lead to the low ap. So i swicthed to find some other loss (such as iouloss and diceloss) to optim the seg branch,but some strange thing appered.  When i trainning serval epochs, the loss change to nan or small number,  the ap is down to 0,  the picture is below,can you give me some advices?

![Image](https://github.com/user-attachments/assets/dc290d3c-aa0a-4afa-b5fb-db670ec391ad)

**Comment by glenn-jocher** [comment]
@luoshiyong the instability and NaN loss you're experiencing when using alternative losses (IoULoss, DiceLoss) for the segmentation branch likely stem from numerical instability or vanishing gradients, especially with small objects. You can try modifying hyperparameters such as learning rate, or explore other loss functions designed to deal with instability, but success is not guaranteed. For details on the `single_mask_loss` function and its parameters, see [Ultralytics Docs](https://docs.ultralytics.com/reference/utils/loss/).
